### PR TITLE
Tree: Integrationtests for filter + selection (#3926)

### LINKED ISSUE #3926: Tree : Multiple selection checkbox propagation inconsistent with filter option
## 1) Environment
- PrimeFaces version: 6.2 
- Does it work on the newest released PrimeFaces version? Version? No
- Does it work on the newest sources in GitHub? (Build by source -> https://github.com/primefaces/primefaces/wiki/Building-From-Source) No
- Application server + version:  Apache Tomcat/8.5.32
- Affected browsers: Firefox, Chrome

## 2) Expected behavior
With or without filtering, I expected to see the same results on selection. Filtering shouldn't change the selected nodes.

image

## 3) Actual behavior
All the parent nodes are selected automatically after filtering and only selecting one single node, not all the children of that node.
image

## 4) Steps to reproduce
Example is below.

## 5) Sample XHTML
    <h:form id="selection">
        <p:growl id="msgs" showDetail="true" escape="false"/>
        <h3>Checkbox</h3>
        <p:tree id="multiCourtTree" value="#{propagationSelectionView.root}" var="_doc"
                    selectionMode="checkbox" propagateSelectionUp="true" propagateSelectionDown="true"
                    filterBy="#{_doc.name}" filterMatchMode="contains"
                    selection="#{propagationSelectionView.selectedNodes}">
            <p:treeNode>
                <h:outputText value="#{_doc.name}"/>
            </p:treeNode>
        </p:tree>
    </h:form>

## 6) Sample bean

public class PropagationSelectionView {

    private TreeNode root;
    private TreeNode[] selectedNodes;

    @PostConstruct
    public void init() {
        root = createCheckboxDocuments();
    }

    public TreeNode createCheckboxDocuments() {
        TreeNode root = new CheckboxTreeNode(new Document("Files"), null);

        TreeNode documents = new CheckboxTreeNode(new Document("Documents"), root);
        TreeNode work = new CheckboxTreeNode(new Document("Work"), documents);
        TreeNode primefaces = new CheckboxTreeNode(new Document("PrimeFaces"), documents);

        //Documents
        TreeNode expenses = new CheckboxTreeNode(new Document("Expenses.doc"), work);
        TreeNode resume = new CheckboxTreeNode( new Document("Resume.doc"), work);
        TreeNode refdoc = new CheckboxTreeNode(new Document("RefDoc.pages"), primefaces);

        return root;
    }

    public TreeNode getRoot() {
        return root;
    }

    public TreeNode[] getSelectedNodes() {
        return selectedNodes;
    }

    public void setSelectedNodes(TreeNode[] selectedNodes) {
        this.selectedNodes = selectedNodes;
    }

    public class Document implements Serializable, Comparable<CheckboxDocument>{

        private String name;

        public Document(String name) {
            this.name = name;
        }

        public String getName() {
            return name;
        }

        public void setName(String name) {
            this.name = name;
        }

        @Override
        public String toString() {
            return name;
        }

        public int compareTo(CheckboxDocument document) {
            return this.getName().compareTo(document.getName());
        }
    }
}

## 7)Conclusion

Selected Nodes should behave in the same way on whole tree structure regardless of filtering.

### PR DESCRIPTION
**Pull request by christophs78**

### CODE REVIEW

File: primefaces-integration-tests/src/test/java/org/primefaces/integrationtests/tree/Tree004Test.java [hunk 1]
```diff
@@ -126,10 +131,132 @@ public void testMultipleSelectionWithoutMetakey(Page page) {
...
+        ComponentUtils.sendKeys(filter, Keys.BACK_SPACE); // null filter press backspace to trigger the re-filtering
+        PrimeSelenium.wait(300 + 100); // 300ms are hardcoded as filterDelay in tree.vertical.js
+        PrimeSelenium.waitGui().until(PrimeExpectedConditions.visibleInViewport(tree.findElement(By.id("form:tree:0"))));
+
+        // Assert Pt. 2
+        page.buttonShowSelectedNodes.click();
+        assertMessage(page.messages, 0, "Selected nodes", "Product Launch");
+        Assertions.assertTrue(PrimeSelenium.hasCssClass(tree.getChildren().get(0).getWebElement().findElement(By.className("ui-chkbox-icon")), "ui-icon-blank"));
+        // TODO: currently broken (2021-09-12)
+        // Assertions.assertTrue(PrimeSelenium.hasCssClass(tree.getChildren().get(1).getWebElement().findElement(By.className("ui-chkbox-icon")), "ui-icon-minus"));
```
Comment (christophs78) [inline primefaces-integration-tests/src/test/java/org/primefaces/integrationtests/tree/Tree004Test.java]: We have one minor issue we may try to fix.

**Comment by melloware** [comment]
Nice work!  @FlipWarthog mentioned we needed more tests for this thanks for jumping on it!

From looking at your test I can't tell exactly what scenario doesn't work can you describe the manual steps and maybe we can look at it?

File: primefaces-integration-tests/src/test/java/org/primefaces/integrationtests/tree/Tree004Test.java [hunk 2]
```diff
@@ -126,10 +131,132 @@ public void testMultipleSelectionWithoutMetakey(Page page) {
...
+    @Test
+    @Order(4)
+    @DisplayName("Tree: Multiple selection + Filter, https://github.com/primefaces/primefaces/issues/3926")
+    public void testFilterWithMultipleSelection(Page page) {
+        // Arrange
+        Tree tree = page.tree;
+        page.selection.select("checkbox");
+        Assertions.assertNotNull(tree);
+
+        List<TreeNode> children = tree.getChildren();
```
Comment (melloware) [inline primefaces-integration-tests/src/test/java/org/primefaces/integrationtests/tree/Tree004Test.java:200]: You have Children as a variable here but but below you repeat yourself with `tree.getChildren()` everywhere making the code harder to read.  Any reason?

Comment (christophs78) [reply primefaces-integration-tests/src/test/java/org/primefaces/integrationtests/tree/Tree004Test.java:200]: Nope. I´ll fix this in the evening.

Comment (christophs78) [reply primefaces-integration-tests/src/test/java/org/primefaces/integrationtests/tree/Tree004Test.java:200]: Had some more minutes to fix this. Now i switch to famliy-time. :-)

**Review by melloware — APPROVED** [review]

**Pull request merged**